File: solutions/2026-07-13/cpp/2026-07-13-sequential-digits.cpp
/*
 * Problem Summary:
 * Generates all integers where each digit is one more than the previous digit,
 * and returns those within a given range [low, high], sorted.
 *
 * Link: https://leetcode.com/problems/sequential-digits/
 *
 * Approach Explanation:
 * The approach enumerates all possible sequential digit numbers by iterating through
 * possible starting digits (1 to 9). For each starting digit, numbers are constructed
 * by successively appending the next digit (e.g., starting with 1, generate 12, then 123, etc.).
 * During construction, if a number falls within the [low, high] range, it's added to a result list.
 * An important optimization is applied: if the `currentNum` being built exceeds `high`,
 * the inner loop breaks, as any further numbers formed by appending more digits will also
 * be greater than `high`.
 * The numbers are generated in a naturally sorted order (first by starting digit, then by length),
 * eliminating the need for an explicit sorting step for the final list.
 *
 * Time Complexity:
 * O(1) - The number of possible sequential digit numbers is very small and constant.
 * There are at most 36 such numbers (e.g., 12, 123, ..., 123456789, 23, 234, ..., 89).
 * The generation process involves a fixed number of outer loop iterations (for starting digits)
 * and inner loop iterations (for appending digits). Each operation within the loops is constant time.
 * Thus, the total time complexity is independent of the input `low` and `high` values
 * beyond their magnitude fitting within standard integer types.
 *
 * Space Complexity:
 * O(1) - The maximum number of sequential digit numbers that can be stored in the result vector
 * is constant (at most 36 integers). The space usage does not scale with `low` or `high`.
 */
#include <vector>     // Required for std::vector
#include <algorithm>  // Not strictly needed for this specific solution as numbers are naturally sorted, but common utility header

class Solution {
public:
    std::vector<int> sequentialDigits(int low, int high) {
        std::vector<int> result; // This vector will store all sequential digit numbers within the range.
        
        // Iterate through all possible starting digits for a sequential number.
        // A sequential number must start with a digit from 1 to 9.
        for (int startDigit = 1; startDigit <= 9; ++startDigit) {
            int currentNum = startDigit;     // Initialize the number with the current starting digit.
            int nextDigit = startDigit + 1;  // Determine the next digit to append.

            // Continue building the sequential number as long as the next digit is valid (1-9).
            while (nextDigit <= 9) {
                // Append the next digit to the current number.
                // Example: If currentNum is 1 and nextDigit is 2, it becomes 12.
                // If currentNum is 12 and nextDigit is 3, it becomes 123.
                currentNum = currentNum * 10 + nextDigit;
                
                // Optimization: If the `currentNum` has already exceeded the `high` bound,
                // any subsequent sequential numbers formed by adding more digits will also
                // definitely be greater than `high`. So, we can stop building numbers
                // for this `startDigit` and move to the next one.
                if (currentNum > high) {
                    break; 
                }
                
                // If the `currentNum` is within the specified range [low, high],
                // add it to our result list.
                if (currentNum >= low) {
                    result.push_back(currentNum);
                }
                
                // Move to the next digit to append in the sequence.
                nextDigit++;
            }
        }
        
        // The numbers are naturally generated in sorted order (e.g., 12, 123, 1234, ..., then 23, 234, ...).
        // Therefore, an explicit std::sort call is not necessary.
        // std::sort(result.begin(), result.end()); // This line is not needed.
        
        return result; // Return the sorted list of sequential digit numbers.
    }
};

File: solutions/2026-07-13/java/2026-07-13-sequential-digits.java
/*
 * Problem Summary:
 * Find all integers within a given range [low, high] where each digit is exactly one greater than the previous digit.
 * The returned list of integers should be sorted in ascending order.
 *
 * Link:
 * https://leetcode.com/problems/sequential-digits/
 *
 * Approach Explanation:
 * The problem involves finding numbers with sequential digits (e.g., 123, 2345, 6789).
 * The total number of such integers is very small (at most 36, ranging from 12 to 123456789), as sequential digits are limited by starting digit and length (cannot exceed 9 digits, and next digit must be <= 9).
 * We can generate all possible sequential digit numbers and then filter them based on the given [low, high] range.
 *
 * The generation strategy involves two nested loops:
 * 1. An outer loop iterates through all possible starting digits (from 1 to 9).
 * 2. An inner loop builds sequential numbers by appending digits. Each appended digit must be one greater than the previous digit (e.g., if the current number being built is 1, append 2 to get 12; then append 3 to get 123, and so on).
 *
 * Inside the inner loop, after forming a new sequential number:
 * - We check if it falls within the [low, high] range. If it does, we add it to our result list.
 * - We also include an optimization: if the `currentNum` has already exceeded `high`, there's no need to continue appending more digits for this sequence. Any further number built from this `currentNum` will also be greater than `high`. Thus, we break out of the inner loop and proceed to the next `startDigit`.
 *
 * After generating all potential sequential numbers, the `result` list is explicitly sorted to meet the problem requirement.
 * Although numbers are generated in increasing order for a given starting digit, and starting digits are processed in increasing order, the overall generation order (e.g., 12, 123, then 23, 234) does not guarantee the final list is sorted without an explicit `Collections.sort()` call.
 *
 * Time Complexity:
 * O(1) - The number of sequential digit numbers possible is fixed and very small (at most 36).
 * The generation process involves constant arithmetic operations within two nested loops, each running at most 9 times.
 * Adding elements to an `ArrayList` takes amortized O(1).
 * Sorting a list of at most 36 elements is also effectively constant time (36 log 36 is a tiny, fixed number of operations).
 * Thus, the total time complexity is independent of the `low` and `high` input values' magnitudes, making it constant.
 *
 * Space Complexity:
 * O(1) - The `result` list stores at most 36 integers. This is a constant amount of memory, not dependent on the input range.
 */
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    public List<Integer> sequentialDigits(int low, int high) {
        // Initialize an ArrayList to store all sequential digit numbers found within the range.
        List<Integer> result = new ArrayList<>();

        // The outer loop iterates through all possible starting digits for a sequential number.
        // A sequential number must start with a digit from 1 to 9.
        for (int startDigit = 1; startDigit <= 9; startDigit++) {
            // 'currentNum' will hold the sequential number being built.
            // It starts with the 'startDigit'.
            int currentNum = startDigit;

            // The inner loop appends subsequent digits to 'currentNum'.
            // Each 'nextDigit' must be one greater than the previous digit used.
            // So, it starts from 'startDigit + 1'.
            for (int nextDigit = startDigit + 1; nextDigit <= 9; nextDigit++) {
                // Append the 'nextDigit' to 'currentNum'.
                // Example: If currentNum is 12 and nextDigit is 3, new currentNum becomes 123.
                currentNum = currentNum * 10 + nextDigit;

                // Check if the newly formed 'currentNum' falls within the specified range [low, high].
                if (currentNum >= low && currentNum <= high) {
                    result.add(currentNum); // If it's in range, add it to our list.
                }

                // Optimization: If 'currentNum' has already exceeded 'high',
                // there's no need to continue appending more digits for this sequence.
                // Any further number built from this 'currentNum' will also be greater than 'high'.
                // Break out of the inner loop and try the next 'startDigit'.
                if (currentNum > high) {
                    break;
                }
            }
        }

        // The problem requires the output list to be sorted.
        // The generation order does not guarantee a fully sorted list across all starting digits
        // (e.g., 12, 123 are generated before 23, 234, but 23 should appear before 123).
        // Therefore, an explicit sort is necessary.
        Collections.sort(result);

        // Return the sorted list of sequential digit numbers within the range.
        return result;
    }
}